File: Deal/src/main/java/ru/leonov/deal/mappers/FinishRegistrationMapper.java
package ru.leonov.deal.mappers;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.factory.Mappers;
import ru.leonov.deal.dto.CreditDTO;
import ru.leonov.deal.dto.FinishRegistrationRequestDTO;
import ru.leonov.deal.dto.ScoringDataDTO;
import ru.leonov.deal.model.entity.ApplicationEntity;
import ru.leonov.deal.model.entity.ClientEntity;
import ru.leonov.deal.model.entity.CreditEntity;

@Mapper
public interface FinishRegistrationMapper {

    FinishRegistrationMapper INSTANCE = Mappers.getMapper(FinishRegistrationMapper.class);

    @Mapping(target = "amount", source = "application.appliedOffer.requestedAmount")
    @Mapping(target = "term", source = "application.appliedOffer.term")
    @Mapping(target = "insuranceEnabled", source = "application.appliedOffer.insuranceEnabled")
    @Mapping(target = "salaryClient", source = "application.appliedOffer.salaryClient")

    @Mapping(target = "firstName", source = "application.client.firstName")
    @Mapping(target = "lastName", source = "application.client.lastName")
    @Mapping(target = "middleName", source = "application.client.middleName")
    @Mapping(target = "birthdate", source = "application.client.birthDate")
    @Mapping(target = "passportSeries", source = "application.client.passport.series")
    @Mapping(target = "passportNumber", source = "application.client.passport.number")
    @Mapping(target = "gender", source = "application.client.gender")
    @Mapping(target = "maritalStatus", source = "application.client.maritalStatus")
    @Mapping(target = "dependentAmount", source = "application.client.dependentAmount")
    @Mapping(target = "passportIssueDate", source = "application.client.passport.issueDate")
    @Mapping(target = "passportIssueBranch", source = "application.client.passport.issueBranch")

    @Mapping(target = "employment.employmentStatus", source = "application.client.employment.employmentStatus")
    @Mapping(target = "employment.employerINN", source = "application.client.employment.employer")
    @Mapping(target = "employment.salary", source = "application.client.employment.salary")
    @Mapping(target = "employment.position", source = "application.client.employment.position")
    @Mapping(target = "employment.workExperienceTotal", source = "application.client.employment.workExperienceTotal")
    @Mapping(target = "employment.workExperienceCurrent", source = "application.client.employment.workExperienceCurrent")

    @Mapping(target = "account", source = "application.client.account")
    ScoringDataDTO mapApplicationEntityToScoringData(ApplicationEntity application);


    @Mapping(target = "additionalServices.isInsuranceEnabled", source = "creditDTO.insuranceEnabled")
    @Mapping(target = "additionalServices.isSalaryClient", source = "creditDTO.salaryClient")
    CreditEntity mapCreditDTOtoCreditEntity(CreditDTO creditDTO);


    @Mapping(target = "passport.series", source = "client.passport.series")
    @Mapping(target = "passport.number", source = "client.passport.number")

    @Mapping(target = "gender", source = "registrationRequest.gender")
    @Mapping(target = "maritalStatus", source = "registrationRequest.maritalStatus")
    @Mapping(target = "dependentAmount", source = "registrationRequest.dependentAmount")
    @Mapping(target = "passport.issueDate", source = "registrationRequest.passportIssueDate")
    @Mapping(target = "passport.issueBranch", source = "registrationRequest.passportIssueBranch")
    @Mapping(target = "employment", source = "registrationRequest.employment")
    @Mapping(target = "employment.employer", source = "registrationRequest.employment.employerINN")
    @Mapping(target = "account", source = "registrationRequest.account")
    ClientEntity mapFinishRegistrationToClient(FinishRegistrationRequestDTO registrationRequest,
                                               ClientEntity client);

}

File: Deal/deal_impl/src/main/kotlin/ru/leonov/deal/mapper/FinishRegistrationMapper.kt
package ru.leonov.deal.mapper

import ru.leonov.deal.dto.CreditDTO
import ru.leonov.deal.dto.EmploymentDTO
import ru.leonov.deal.dto.FinishRegistrationRequestDTO
import ru.leonov.deal.dto.ScoringDataDTO
import ru.leonov.deal.model.entity.ApplicationEntity
import ru.leonov.deal.model.entity.ClientEntity
import ru.leonov.deal.model.entity.CreditEntity
import ru.leonov.deal.model.enums.ClientGenderEnum
import ru.leonov.deal.model.enums.ClientMartialStatusEnum
import ru.leonov.deal.model.enums.EmploymentPositionEnum
import ru.leonov.deal.model.enums.EmploymentStatusEnum
import ru.leonov.deal.model.record.AdditionalServicesRecord
import ru.leonov.deal.model.record.EmploymentRecord
import ru.leonov.deal.model.record.PassportRecord

/**
 * Create new [ScoringDataDTO] object from [ApplicationEntity] data.
 * @throws IllegalArgumentException if required data field in [ApplicationEntity] is null.
 */
fun mapApplicationEntityToScoringDataDTO(application: ApplicationEntity): ScoringDataDTO {

    val employmentRecord = application.client.employment
        ?: throw IllegalArgumentException("Can't map scoring data: employment is null!")

    val employmentDto = EmploymentDTO(
        employmentStatus = EmploymentDTO.EmploymentStatus.valueOf(employmentRecord.employmentStatus.name),
        employerINN = employmentRecord.employer,
        salary = employmentRecord.salary,
        position = EmploymentDTO.Position.valueOf(employmentRecord.position.name),
        workExperienceTotal = employmentRecord.workExperienceTotal,
        workExperienceCurrent = employmentRecord.workExperienceCurrent,
    )

    val loanOfferDTO = application.appliedOffer
        ?: throw IllegalArgumentException("Can't map scoring data: appliedOffer is null!")

    return ScoringDataDTO(
        amount = loanOfferDTO.requestedAmount,
        term = loanOfferDTO.term,
        isInsuranceEnabled = loanOfferDTO.isInsuranceEnabled,
        isSalaryClient = loanOfferDTO.isSalaryClient,
        firstName = application.client.firstName,
        lastName = application.client.lastName,
        middleName = application.client.middleName,
        birthdate = application.client.birthDate,
        passportSeries = application.client.passport.series.toString(),
        passportNumber = application.client.passport.number.toString(),
        gender = ScoringDataDTO.Gender.valueOf(
            application.client.gender?.name
                ?: throw IllegalArgumentException("Can't map scoring data: gender is null!")
        ),
        maritalStatus = ScoringDataDTO.MaritalStatus.valueOf(
            application.client.maritalStatus?.name
                ?: throw IllegalArgumentException("Can't map scoring data: martialStatus is null!")
        ),
        dependentAmount = application.client.dependentAmount
            ?: throw IllegalArgumentException("Can't map scoring data: dependentAmount is null!"),
        passportIssueDate = application.client.passport.issueDate
            ?: throw IllegalArgumentException("Can't map scoring data: passportIssueDate is null!"),
        passportIssueBranch = application.client.passport.issueBranch
            ?: throw IllegalArgumentException("Can't map scoring data: passportIssueBranch is null!"),
        account = application.client.account?.toString()
            ?: throw IllegalArgumentException("Can't map scoring data: account is null!"),
        employment = employmentDto
    )
}

/**
 * Create new [CreditEntity] from given [CreditDTO].
 */
fun mapCreditDTOtoCreditEntity(creditDTO: CreditDTO): CreditEntity {

    val additionalServicesRecord = AdditionalServicesRecord(
        isInsuranceEnabled = creditDTO.isInsuranceEnabled,
        isSalaryClient = creditDTO.isSalaryClient
    )

    return CreditEntity(
        amount = creditDTO.amount,
        term = creditDTO.term,
        monthlyPayment = creditDTO.monthlyPayment,
        rate = creditDTO.rate,
        psk = creditDTO.psk,
        paymentSchedule = creditDTO.paymentSchedule,
        additionalServices = additionalServicesRecord
    )
}

/**
 * Create new completed [ClientEntity] based on incomplete [ClientEntity] and [FinishRegistrationRequestDTO].
 */
fun mapFinishRegistrationToClient(registrationRequest: FinishRegistrationRequestDTO, client: ClientEntity):
        ClientEntity {

    val passportRecord = PassportRecord(
        series = client.passport.series,
        number = client.passport.number,
        issueDate = registrationRequest.passportIssueDate,
        issueBranch = registrationRequest.passportIssueBranch
    )

    val employmentRecord = EmploymentRecord(
        employmentStatus = EmploymentStatusEnum.valueOf(registrationRequest.employment.employmentStatus.value),
        employer = registrationRequest.employment.employerINN,
        salary = registrationRequest.employment.salary,
        position = EmploymentPositionEnum.valueOf(registrationRequest.employment.position.value),
        workExperienceTotal = registrationRequest.employment.workExperienceTotal,
        workExperienceCurrent = registrationRequest.employment.workExperienceCurrent
    )

    return ClientEntity(

        firstName = client.firstName,
        lastName = client.lastName,
        middleName = client.middleName,
        birthDate = client.birthDate,
        email = client.email,
        gender = ClientGenderEnum.valueOf(registrationRequest.gender.value),
        id = client.id,

        passport = passportRecord,
        employment = employmentRecord,
        maritalStatus = ClientMartialStatusEnum.valueOf(registrationRequest.maritalStatus.value),
        dependentAmount = registrationRequest.dependentAmount,
        account = registrationRequest.account.toLong()
    )
}
